pkg/etcd/etcd.go

```diff
@@ -505,46 +522,139 @@ func (e *ETCD) removePeer(ctx context.Context, id, address string) error {
-			if !member.IsLearner {
-				continue
-			}
-			if _, err := client.MemberPromote(ctx, member.ID); err != nil {
-				memberPromoted = false
+			if member.IsLearner {
+				if err := e.trackLearnerProgress(ctx, progress, member); err != nil {
+					logrus.Errorf("Failed to track learner progress towards promotion: %v", err)
+				}
 				break
 			}
 		}
-		if memberPromoted {
-			break
-		}
 	}
 	return nil
 }
 
+// trackLearnerProcess attempts to promote a learner. If it cannot be promoted, progress through the raft index is tracked.
+// If the learner does not make any progress in a reasonable amount of time, it is evicted from the cluster.
+func (e *ETCD) trackLearnerProgress(ctx context.Context, progress *learnerProgress, member *etcdserverpb.Member) error {
+	// Try to promote it. If it can be promoted, no further tracking is necessary
+	if _, err := e.client.MemberPromote(ctx, member.ID); err != nil {
+		logrus.Debugf("Unable to promote learner %s: %v", member.Name, err)
+	} else {
+		logrus.Infof("Promoted learner %s", member.Name)
+		return nil
+	}
+
+	now := time.Now()
+
+	// If this is the first time we've tracked this member's progress, reset stats
+	if progress.Name != member.Name || progress.ID != member.ID {
+		progress.ID = member.ID
+		progress.Name = member.Name
+		progress.RaftAppliedIndex = 0
+		progress.LastProgress.Time = now
+	}
+
+	// Update progress by retrieving status from the member's first reachable client URL
+	for _, ep := range member.ClientURLs {
+		ctx, cancel := context.WithTimeout(ctx, defaultDialTimeout)
+		defer cancel()
+		status, err := e.client.Status(ctx, ep)
+		if err != nil {
+			logrus.Debugf("Failed to get etcd status from learner %s at %s: %v", member.Name, ep, err)
+			continue
+		}
+
+		if progress.RaftAppliedIndex < status.RaftAppliedIndex {
+			logrus.Debugf("Learner %s has progressed from RaftAppliedIndex %d to %d", progress.Name, progress.RaftAppliedIndex, status.RaftAppliedIndex)
+			progress.RaftAppliedIndex = status.RaftAppliedIndex
+			progress.LastProgress.Time = now
+		}
+		break
+	}
+
+	// Warn if the learner hasn't made any progress
+	if !progress.LastProgress.Time.Equal(now) {
+		logrus.Warnf("Learner %s stalled at RaftAppliedIndex=%d for %s", progress.Name, progress.RaftAppliedIndex, now.Sub(progress.LastProgress.Time).String())
+	}
+
+	// See if it's time to evict yet
+	if now.Sub(progress.LastProgress.Time) > learnerMaxStallTime {
+		if _, err := e.client.MemberRemove(ctx, member.ID); err != nil {
+			return err
+		}
+		logrus.Warnf("Removed learner %s from etcd cluster", member.Name)
+		return nil
+	}
+
+	return e.setLearnerProgress(ctx, progress)
+}
+
+// getLearnerProgress returns the stored learnerProgress struct as retrieved from etcd
+func (e *ETCD) getLearnerProgress(ctx context.Context) (*learnerProgress, error) {
+	progress := &learnerProgress{}
+
+	value, err := e.client.Get(ctx, learnerProgressKey)
+	if err != nil {
+		if err != rpctypes.ErrGRPCKeyNotFound {
```

Comment:
for some reason this didnt work for me, I am curious if that worked for you?

Reply:
It looks like this is a v2 error - v3 just returns a response with 0 key/values which is already handled below.